### PR TITLE
Add cluster stats re. snapshot activity

### PR DESCRIPTION
**Pull request by DaveCTurner**
Shows how many ongoing snapshots/clones/deletions/etc. there are, and summarises the shard-level status too for progress tracking.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/cluster/ClusterSnapshotStats.java [hunk 1]
```diff
@@ -0,0 +1,269 @@
...
+        int snapshotCount,
+        int cloneCount,
+        int finalizationsCount,
+        int totalShards,
+        int completeShards,
+        int deletionsCount,
+        int snapshotDeletionsCount,
+        int activeDeletionsCount,
+        long firstStartTimeMillis,
+        EnumMap<SnapshotsInProgress.ShardState, Integer> shardStates
```
Comment (idegtiarenko) [inline server/src/main/java/org/elasticsearch/cluster/ClusterSnapshotStats.java:202]: I wonder if this should be a `Map<String, String>` instead.

Writing and reading to stream treat keys as strings anyways.
Building json could be simplified by using `XContentBuilder#field(String name, Map<String, Object> values)`.

Map is copied anyways when creating PerRepositoryStats record.

File: server/src/main/java/org/elasticsearch/cluster/ClusterSnapshotStats.java [hunk 2]
```diff
@@ -0,0 +1,269 @@
...
+
+            var snapshotCount = 0;
+            var cloneCount = 0;
+            var finalizationsCount = 0;
+            var totalShards = 0;
+            var completeShards = 0;
+            final var shardStatesAccumulator = new EnumMap<>(
+                Arrays.stream(SnapshotsInProgress.ShardState.values())
+                    .collect(Collectors.toMap(Function.identity(), ignored -> new AtomicInteger()))
+            );
```
Comment (idegtiarenko) [inline server/src/main/java/org/elasticsearch/cluster/ClusterSnapshotStats.java:77]: NIT: I think it should be possible to use overloaded toMap version that has a map supplier to avoid copying

File: server/src/main/java/org/elasticsearch/cluster/ClusterSnapshotStats.java (hunk 1 above)
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/cluster/ClusterSnapshotStats.java:202]: The main reason I prefer `EnumMap` here is that it orders the keys according to the underlying `enum`. We could achieve the same thing with `LinkedHashMap` everywhere instead but this is easy to get wrong.

File: server/src/main/java/org/elasticsearch/cluster/ClusterSnapshotStats.java (hunk 2 above)
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/cluster/ClusterSnapshotStats.java]: ah yeah so it is

File: docs/reference/cluster/stats.asciidoc [hunk 3]
```diff
@@ -1307,6 +1307,125 @@ Number of selected nodes using the distribution flavor and file type.
...
+=====
+
+`current_counts`:::
+(object)
+Contains statistics which report the numbers of various ongoing snapshot activities in the cluster.
++
+.Properties of `current_counts`
+[%collapsible%open]
+======
+`snapshots`:::
```
Comment (kingherc) [inline docs/reference/cluster/stats.asciidoc:1325]: I wonder if it'd make sense to extend also the `_cat/snapshots/<repository>` API with any new of these counts that you think may be significant enough to be easily accessible there?

**Review by kingherc — COMMENTED** [review]
Nice one! Some comments/questions

File: test/framework/src/main/java/org/elasticsearch/test/ESTestCase.java [hunk 4]
```diff
@@ -815,6 +815,10 @@ public static IntStream randomInts(long streamSize) {
         return random().ints(streamSize);
     }
 
+    public static int randomNonNegativeInt() {
+        return randomInt() & Integer.MAX_VALUE;
```
[the remark this answers is not in the document]
Comment (DaveCTurner) [reply test/framework/src/main/java/org/elasticsearch/test/ESTestCase.java:819]: Yeah ok see https://github.com/elastic/elasticsearch/pull/93705

File: docs/reference/cluster/stats.asciidoc (hunk 3 above)
Comment (DaveCTurner) [reply docs/reference/cluster/stats.asciidoc:1325]: I think you already get enough useful detail in this API - for ongoing snapshots you can see the shard-level progress. But this API shows one row per snapshot so it's not obvious to me how you'd show info about ongoing deletions or clones there.

In principle we could add it to `GET _cat/repositories` I suppose, but I think it's not worth the effort right now.

File: server/src/main/java/org/elasticsearch/cluster/ClusterSnapshotStats.java [hunk 5]
```diff
@@ -0,0 +1,270 @@
...
+        out.writeVLong(incompleteShardSnapshotCount);
+        out.writeVInt(deletionsInProgressCount);
+        out.writeVInt(cleanupsInProgressCount);
+        out.writeList(statsByRepository);
+    }
+
+    public static ClusterSnapshotStats readFrom(StreamInput in) throws IOException {
+        return new ClusterSnapshotStats(
+            in.readVInt(),
+            in.readVInt(),
```
[the remark this answers is not in the document]
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/cluster/ClusterSnapshotStats.java:184]: Oof well spotted.

File: docs/reference/cluster/stats.asciidoc [hunk 6]
```diff
@@ -1307,6 +1307,125 @@ Number of selected nodes using the distribution flavor and file type.
...
+(integer)
+The total number of snapshot deletion operations currently in this repository.
+
+`snapshot_deletions`:::
+(integer)
+The total number of snapshots that are currently being deleted from this repository.
+
+`active_deletions`:::
+(integer)
+The total number of snapshot deletion operations currently active in this repository.
```
[the remark this answers is not in the document]
Comment (DaveCTurner) [reply docs/reference/cluster/stats.asciidoc:1386]: :) I was hoping nobody would ask. This starts to get into the details of how the snapshot state machine works, which isn't documented today. I think this number might be useful for debugging but I'd rather not have to describe the implementation deeply enough to explain it.

TLDR we don't delete snapshots and create snapshots concurrently, and while we're busy creating snapshots this will be `0` even if there are deletes waiting to run.

File: docs/reference/cluster/stats.asciidoc [hunk 7]
```diff
@@ -1307,6 +1307,125 @@ Number of selected nodes using the distribution flavor and file type.
...
+The total number of tracked shard snapshots which have completed in this repository.
+
+`incomplete`:::
+(integer)
+The total number of tracked shard snapshots which have not completed in this repository. This is the difference between the
+`total` and `complete` values.
+
+`states`:::
+(object)
+The total number of shard snapshots in each of the named states in this repository.
```
[the remark this answers is not in the document]
Comment (DaveCTurner) [reply docs/reference/cluster/stats.asciidoc:1411]: Again, IMO this feels like overly-deep implementation details.

File: docs/reference/cluster/stats.asciidoc (hunk 6 above)
[the remark this answers is not in the document]
Comment (kingherc) [reply docs/reference/cluster/stats.asciidoc:1386]: I guess you can add `(useful for debugging)` just so as to avoid questions :D

File: docs/reference/cluster/stats.asciidoc [hunk 8]
```diff
@@ -1307,6 +1307,125 @@ Number of selected nodes using the distribution flavor and file type.
...
++
+.Properties of `snapshots`
+[%collapsible%open]
+=====
+
+`current_counts`:::
+(object)
+Contains statistics which report the numbers of various ongoing snapshot activities in the cluster.
++
+.Properties of `current_counts`
```
[the remark this answers is not in the document]
Comment (DaveCTurner) [reply docs/reference/cluster/stats.asciidoc:1322]: I want to say that they're `counts` too here, rather than calling the fields `snapshots_count`, `shard_snapshots_count`, and so on. Moreover we use `current` in this way in other stats APIs.

**Review by kingherc — APPROVED** [review]
LGTM. My remaining comments are optional

File: docs/reference/cluster/stats.asciidoc [hunk 9]
```diff
@@ -1307,6 +1307,125 @@ Number of selected nodes using the distribution flavor and file type.
...
+(integer)
+The total number of snapshots that are currently being deleted from this repository.
+
+`active_deletions`:::
+(integer)
+The total number of snapshot deletion operations currently active in this repository.
+
+`shards`:::
+(object)
+Contains statistics which report the shard-level progress of snapshot activities for a repository.
```
[the remark this answers is not in the document]
Comment (DaveCTurner) [reply docs/reference/cluster/stats.asciidoc]: The key bit here is the word `tracked`. We only track shard counts for ongoing snapshots. I do agree it's a little confusing tho, I wrote some more words in  dd41889d0ca3cfae7bf71050b285dabe7eeef8b0.
